services/motion/localizer.go

```diff
@@ -2,59 +2,70 @@ package motion
 type Localizer interface {
-	CurrentPosition(context.Context) (referenceframe.PoseInFrame, error)
-}
-
-// NewLocalizer constructs either a slamLocalizer or movementSensorLocalizer from the given resource.
-func NewLocalizer(ctx context.Context, res resource.Resource) (Localizer, error) {
-	switch res := res.(type) {
-	case slam.Service:
-		return &slamLocalizer{Service: res}, nil
-	case movementsensor.MovementSensor:
-		return &movementSensorLocalizer{MovementSensor: res}, nil
-	default:
-		return nil, fmt.Errorf("cannot localize on resource of type %T", res)
-	}
+	CurrentPosition(context.Context) (*referenceframe.PoseInFrame, error)
 }
 
 // slamLocalizer is a struct which only wraps an existing slam service.
 type slamLocalizer struct {
 	slam.Service
 }
 
+func NewSLAMLocalizer(slam slam.Service) Localizer {
+	return &slamLocalizer{Service: slam}
+}
+
 // CurrentPosition returns slam's current position.
-func (s slamLocalizer) CurrentPosition(ctx context.Context) (referenceframe.PoseInFrame, error) {
-	var pif referenceframe.PoseInFrame
+func (s *slamLocalizer) CurrentPosition(ctx context.Context) (*referenceframe.PoseInFrame, error) {
 	pose, _, err := s.GetPosition(ctx)
 	if err != nil {
-		return pif, err
+		return nil, err
 	}
-	return *referenceframe.NewPoseInFrame(referenceframe.World, pose), err
+	return referenceframe.NewPoseInFrame(referenceframe.World, pose), err
 }
 
 // movementSensorLocalizer is a struct which only wraps an existing movementsensor.
 type movementSensorLocalizer struct {
 	movementsensor.MovementSensor
+	origin      *geo.Point
+	calibration spatialmath.Pose
+}
+
+func NewMovementSensorLocalizer(ms movementsensor.MovementSensor, origin *geo.Point, calibration spatialmath.Pose) Localizer {
+	return &movementSensorLocalizer{MovementSensor: ms, origin: origin, calibration: calibration}
 }
 
 // CurrentPosition returns a movementsensor's current position.
-func (m movementSensorLocalizer) CurrentPosition(ctx context.Context) (referenceframe.PoseInFrame, error) {
-	var pif referenceframe.PoseInFrame
+func (m *movementSensorLocalizer) CurrentPosition(ctx context.Context) (*referenceframe.PoseInFrame, error) {
 	gp, _, err := m.Position(ctx, nil)
 	if err != nil {
-		return pif, err
+		return nil, err
+	}
+	var o spatialmath.Orientation
+	compass, err := m.CompassHeading(ctx, nil)
+	if err != nil {
+		if !errors.Is(err, movementsensor.ErrMethodUnimplementedCompassHeading) {
+			return nil, err
+		}
+		o, err = m.Orientation(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		o = &spatialmath.OrientationVectorDegrees{OZ: 1, Theta: compass}
 	}
-	pose := spatialmath.GeoPointToPose(gp)
-	return *referenceframe.NewPoseInFrame(m.Name().Name, pose), nil
+	pose := spatialmath.NewPose(spatialmath.GeoPointToPose(gp, m.origin).Point(), o)
+	correction := spatialmath.Compose(m.calibration, spatialmath.NewPoseFromOrientation(&spatialmath.OrientationVector{OZ: 1, Theta: -math.Pi / 2}))
```

Comment:
Both this and the Compose below are very unintuitive to follow, and will go away if the calibration is done in the FS as mentioned above.